### PR TITLE
[engine] stack-safe alpha equivalence

### PR DESCRIPTION
**Pull request by nickchapman-da**
Make alpha equivalence be stack-safe:
- [x] new test-cases to demonstrate stack-overflow
- [x] fix stack-overflow

This advances https://github.com/digital-asset/daml/issues/13410

### CODE REVIEW

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/AlphaEquiv.scala [hunk 1]
```diff
@@ -5,39 +5,59 @@ package com.daml.lf.validation
...
+          (fs1.names sameElements fs2.names) && {
+            val more = (fs1.values zip fs2.values).toList.map { case (x1, x2) => (env, x1, x2) }
+            alphaEquivList(more ++ trips)
+          }
+        case (TSynApp(f, xs), TSynApp(g, ys)) =>
+          // We treat type synonyms nominally here. If alpha equivalence
+          // fails, we expand all of them and try again.
+          f == g && xs.length == ys.length && {
+            val more = (xs.iterator zip ys.iterator).toList.map { case (x1, x2) => (env, x1, x2) }
+            alphaEquivList(more ++ trips)
```
Comment (remyhaemmerle-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/AlphaEquiv.scala:57]:
```suggestion
            val more = (xs.iterator zip ys.iterator).map { case (x1, x2) => (env, x1, x2) }
            alphaEquivList(more ++: trips)
```

**Review by remyhaemmerle-da — APPROVED** [review]
Nice.
Could you run some benchmark for the type checker (more for information purpose)? I think remember the alpha equivalence was one of the hot spot.

**Comment by nickchapman-da** [comment]
type checker benchmark.. differences dwarfed by the noise:

HEAD...
```
  Result "com.daml.lf.validation.TypecheckingBench.bench":
    5.841 ±(99.9%) 0.171 s/op [Average]
    (min, avg, max) = (5.786, 5.841, 5.879), stdev = 0.044
    CI (99.9%): [5.669, 6.012] (assumes normal distribution)
```
BASE...
```
  Result "com.daml.lf.validation.TypecheckingBench.bench":
    5.903 ±(99.9%) 0.539 s/op [Average]
    (min, avg, max) = (5.733, 5.903, 6.117), stdev = 0.140
    CI (99.9%): [5.364, 6.441] (assumes normal distribution)
```

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/AlphaEquiv.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/AlphaEquiv.scala]: I googled for info about this operator:
> ++ and ++: return different results when the operands are different types of collection. ++ returns the same collection type as the left side, and ++: returns the same collection type as the right side

In this specific case, both sides were already lists, so I think it was ok to use `++`

But I switched to `++:` as you suggest, and this does seem to be a better default choice.

**Pull request merged**